bin/sam-coverage-depth.py

```diff
@@ -2,169 +2,391 @@
+        if referenceId not in references:
+            raise ValueError(
+                f"Reference id {referenceId} not found in {filename!r} "
+                f"Known references are {', '.join(sorted(references))}."
+            )
+    else:
+        if len(references) == 1:
+            referenceId = list(references)[0]
+        else:
+            raise ValueError(
+                f"Multiple references found in {filename!r}: "
+                f"{', '.join(sorted(references))}. Use --referenceId "
+                "to specify the one you want."
+            )
 
-    try:
-        referenceLengths = samFilter.referenceLengths()
-    except UnknownReference:
-        referenceId = samFilter.referenceIds.pop()
-        referenceIds = samReferences(args.samfile)
-        print(
-            "Reference %r does not appear in SAM file %s. Known "
-            "references are: %s."
-            % (referenceId, args.samfile, ", ".join(sorted(referenceIds))),
-            file=sys.stderr,
-        )
-        sys.exit(1)
+    return referenceId
 
-    if args.filter:
 
-        def filterRead(read):
-            return not (read.is_del or read.is_refskip) and samFilter.filterAlignment(
-                read.alignment
+def getReferenceSeq(fasta: str | None, referenceId: str) -> str | None:
+    """
+    Look for the reference id in the FASTA file and return its sequence if found.
+    If no FASTA is given, return None.
+    """
+    if fasta:
+        for read in FastaReads(fasta):
+            if read.id == referenceId:
+                return read.sequence
+        else:
+            raise ValueError(
+                f"Reference id {referenceId!r} was not found in reference FASTA "
+                f"file {fasta!r}."
             )
-
     else:
-        # Do not do our custom SAM filtering.
-        def filterRead(read):
-            return not (read.is_del or read.is_refskip)
+        return None
 
-    if args.printStats:
-        counts = []
 
-    with samfile(args.samfile) as sam:
-        if samFilter.referenceIds:
-            # No need to check if the given reference id is in referenceLengths
-            # because the samFilter.referenceLengths call above catches that.
-            referenceId = samFilter.referenceIds.pop()
-        else:
-            if len(referenceLengths) == 1:
-                referenceId = list(referenceLengths)[0]
-            else:
-                print(
-                    "SAM file %r contains %d references (%s). Only one "
-                    "reference id can be analyzed at a time. Please use "
-                    "--referenceId to specify the one you want examined."
-                    % (
-                        args.samfile,
-                        len(referenceLengths),
-                        ", ".join(sorted(referenceLengths)),
-                    ),
-                    file=sys.stderr,
-                )
-                sys.exit(1)
+def processSubsection(
+    args: argparse.Namespace,
```

Comment:
I don't like this, it forces me to look through the function body to figure out what the actual arguments are and it makes meaningful type annotations impossible for those args.

Reply:
OK, agreed & fixed.